Investigate possible push replication issue

1.4-39, ForestDB+Encrytion sg=1.2.1-4 -
test_replication_with_multiple_client_dbs_and_single_sync_gateway_db - http://uberjenkins.sc.couchbase.com/job/android-listener-client-sync-gateway/177/testReport/testsuites.listener.shared.client_sg/test_replication/test_replication_with_multiple_client_dbs_and_single_sync_gateway_db/

Note to Self:

One of failing scenario is to fail to push document. 
```
PUT 1000 docs to http://192.168.57.101:60000/ls_db1/ with prefix ls_db1


―――――――――――――――――――――――――――――――――――――――――― test_replication_with_multiple_client_dbs_and_single_sync_gateway_db ―――――――――――――――――――――――――――――――――――――――――――

setup_client_syncgateway_test = {'cluster_config': 'resources/cluster_configs/base_cc', 'ls_url': 'http://192.168.57.101:60000', 'sg_admin_url': 'http://192.168.33.11:4985', 'sg_mode': 'cc', ...}

    @pytest.mark.sanity
    @pytest.mark.listener
    @pytest.mark.syncgateway
    @pytest.mark.replication
    def test_replication_with_multiple_client_dbs_and_single_sync_gateway_db(setup_client_syncgateway_test):
        """Test replication from multiple client dbs to one sync_gateway db"""
    
        cluster_config = setup_client_syncgateway_test["cluster_config"]
        sg_mode = setup_client_syncgateway_test["sg_mode"]
        ls_url = setup_client_syncgateway_test["ls_url"]
        sg_url = setup_client_syncgateway_test["sg_url"]
        sg_admin_url = setup_client_syncgateway_test["sg_admin_url"]
    
        num_docs = 1000
    
        sg_config = sync_gateway_config_path_for_mode("listener_tests/listener_tests", sg_mode)
        c = cluster.Cluster(config=cluster_config)
        c.reset(sg_config_path=sg_config)
    
        log_info("Running 'test_replication_with_multiple_client_dbs_and_single_sync_gateway_db'")
        log_info("ls_url: {}".format(ls_url))
        log_info("sg_admin_url: {}".format(sg_admin_url))
        log_info("sg_url: {}".format(sg_url))
    
        client = MobileRestClient()
    
        sg_db = "db"
        ls_db1 = client.create_database(url=ls_url, name="ls_db1")
        ls_db2 = client.create_database(url=ls_url, name="ls_db2")
    
        # Setup continuous push / pull replication from ls_db1 to sg_db
        client.start_replication(
            url=ls_url,
            continuous=True,
            from_db=ls_db1,
            to_url=sg_admin_url, to_db=sg_db
        )
    
        client.start_replication(
            url=ls_url,
            continuous=True,
            from_url=sg_admin_url, from_db=sg_db,
            to_db=ls_db1
        )
    
        # Setup continuous push / pull replication from ls_db2 to sg_db
        client.start_replication(
            url=ls_url,
            continuous=True,
            from_db=ls_db2,
            to_url=sg_admin_url, to_db=sg_db
        )
    
        client.start_replication(
            url=ls_url,
            continuous=True,
            from_url=sg_admin_url, from_db=sg_db,
            to_db=ls_db2
        )
    
>       ls_db_one_docs = client.add_docs(url=ls_url, db=ls_db1, number=num_docs, id_prefix=ls_db1)

testsuites/listener/shared/client_sg/test_replication.py:949: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
keywords/MobileRestClient.py:1111: in add_docs
    resp = self._session.put("{}/{}/{}".format(url, db, doc_id), data=data)
venv/lib/python2.7/site-packages/requests/sessions.py:521: in put
    return self.request('PUT', url, data=data, **kwargs)
venv/lib/python2.7/site-packages/requests/sessions.py:468: in request
    resp = self.send(prep, **send_kwargs)
venv/lib/python2.7/site-packages/requests/sessions.py:576: in send
    r = adapter.send(request, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <requests.adapters.HTTPAdapter object at 0x10ab6ec50>, request = <PreparedRequest [PUT]>, stream = False
timeout = <requests.packages.urllib3.util.timeout.Timeout object at 0x109818a50>, verify = True, cert = None, proxies = OrderedDict()

    def send(self, request, stream=False, timeout=None, verify=True, cert=None, proxies=None):
        """Sends PreparedRequest object. Returns Response object.
    
            :param request: The :class:`PreparedRequest <PreparedRequest>` being sent.
            :param stream: (optional) Whether to stream the request content.
            :param timeout: (optional) How long to wait for the server to send
                data before giving up, as a float, or a :ref:`(connect timeout,
                read timeout) <timeouts>` tuple.
            :type timeout: float or tuple
            :param verify: (optional) Whether to verify SSL certificates.
            :param cert: (optional) Any user-provided SSL certificate to be trusted.
            :param proxies: (optional) The proxies dictionary to apply to the request.
            """
    
        conn = self.get_connection(request.url, proxies)
    
        self.cert_verify(conn, request.url, verify, cert)
        url = self.request_url(request, proxies)
        self.add_headers(request)
    
        chunked = not (request.body is None or 'Content-Length' in request.headers)
    
        if isinstance(timeout, tuple):
            try:
                connect, read = timeout
                timeout = TimeoutSauce(connect=connect, read=read)
            except ValueError as e:
                # this may raise a string formatting error.
                err = ("Invalid timeout {0}. Pass a (connect, read) "
                       "timeout tuple, or a single float to set "
                       "both timeouts to the same value".format(timeout))
                raise ValueError(err)
        else:
            timeout = TimeoutSauce(connect=timeout, read=timeout)
    
        try:
            if not chunked:
                resp = conn.urlopen(
                    method=request.method,
                    url=url,
                    body=request.body,
                    headers=request.headers,
                    redirect=False,
                    assert_same_host=False,
                    preload_content=False,
                    decode_content=False,
                    retries=self.max_retries,
                    timeout=timeout
                )
    
            # Send the request.
            else:
                if hasattr(conn, 'proxy_pool'):
                    conn = conn.proxy_pool
    
                low_conn = conn._get_conn(timeout=DEFAULT_POOL_TIMEOUT)
    
                try:
                    low_conn.putrequest(request.method,
                                        url,
                                        skip_accept_encoding=True)
    
                    for header, value in request.headers.items():
                        low_conn.putheader(header, value)
    
                    low_conn.endheaders()
    
                    for i in request.body:
                        low_conn.send(hex(len(i))[2:].encode('utf-8'))
                        low_conn.send(b'\r\n')
                        low_conn.send(i)
                        low_conn.send(b'\r\n')
                    low_conn.send(b'0\r\n\r\n')
    
                    # Receive the response from the server
                    try:
                        # For Python 2.7+ versions, use buffering of HTTP
                        # responses
                        r = low_conn.getresponse(buffering=True)
                    except TypeError:
                        # For compatibility with Python 2.6 versions and back
                        r = low_conn.getresponse()
    
                    resp = HTTPResponse.from_httplib(
                        r,
                        pool=conn,
                        connection=low_conn,
                        preload_content=False,
                        decode_content=False
                    )
                except:
                    # If we hit any problems here, clean up the connection.
                    # Then, reraise so that we can handle the actual exception.
                    low_conn.close()
                    raise
    
        except (ProtocolError, socket.error) as err:
>           raise ConnectionError(err, request=request)
E           ConnectionError: ('Connection aborted.', error(54, 'Connection reset by peer'))

venv/lib/python2.7/site-packages/requests/adapters.py:426: ConnectionError
```

Attachment indicates `PUT http://192.168.57.101:60000/ls_db1/ls_db1_990` request reset by peer. 
![screen shot 2017-02-09 at 12 52 33 pm](https://cloud.githubusercontent.com/assets/15889/22802684/a4dda95e-eec6-11e6-90a3-e2a3d7c1fde2.png)

Request Data
```
00000000  7b 22 6c 69 73 74 5f 6f 66 5f 64 69 63 74 73 22 3a 20 5b 7b 22 66 72 69 65 6e  {"list_of_dicts": [{"frien
0000001a  64 5f 6f 6e 65 22 3a 20 22 78 4f 79 65 76 65 70 57 57 4d 22 7d 2c 20 7b 22 66  d_one": "xOyevepWWM"}, {"f
00000034  72 69 65 6e 64 5f 74 77 6f 22 3a 20 22 51 6f 5a 4d 46 63 5a 78 6b 62 22 7d 5d  riend_two": "QoZMFcZxkb"}]
0000004e  2c 20 22 64 61 74 65 5f 74 69 6d 65 5f 61 64 64 65 64 22 3a 20 22 32 30 31 37  , "date_time_added": "2017
00000068  2d 30 32 2d 30 39 20 31 32 3a 34 39 3a 30 36 2e 34 34 37 33 36 36 22 2c 20 22  -02-09 12:49:06.447366", "
00000082  64 69 63 74 5f 77 69 74 68 5f 6c 69 73 74 22 3a 20 7b 22 6c 69 73 74 22 3a 20  dict_with_list": {"list": 
0000009c  5b 66 61 6c 73 65 2c 20 74 72 75 65 5d 7d 2c 20 22 6c 69 73 74 22 3a 20 5b 31  [false, true]}, "list": [1
000000b6  39 38 35 38 33 31 32 31 30 38 30 34 35 36 35 36 30 33 2c 20 34 37 31 38 32 37  985831210804565603, 471827
000000d0  37 38 39 35 36 37 37 38 36 33 38 35 30 5d 2c 20 22 64 69 63 74 22 3a 20 7b 22  7895677863850], "dict": {"
000000ea  6e 61 6d 65 22 3a 20 22 4e 55 44 44 58 6c 4f 52 61 59 22 7d 2c 20 22 75 70 64  name": "NUDDXlORaY"}, "upd
00000104  61 74 65 73 22 3a 20 30 7d                                                     ates": 0} 
```

Unfortunately, I can not found any log info about `PUT /ls_db1/ls_db1_990/` from logcat.
Full LogCat
https://gist.github.com/hideki/aec6a1c463f243338fc7910dd577fff7

```
>>> Running in channel cache
Running 'test_replication_with_multiple_client_dbs_and_single_sync_gateway_db'
ls_url: http://192.168.57.101:60000
sg_admin_url: http://192.168.33.11:4985
sg_url: http://192.168.33.11:4984
GET http://192.168.57.101:60000/ 200
PUT http://192.168.57.101:60000/ls_db1/ 201
GET http://192.168.57.101:60000/ls_db1/ 200
GET http://192.168.57.101:60000/ 200
PUT http://192.168.57.101:60000/ls_db2/ 201
GET http://192.168.57.101:60000/ls_db2/ 200
POST http://192.168.57.101:60000/_replicate 200
POST http://192.168.57.101:60000/_replicate 200
POST http://192.168.57.101:60000/_replicate 200
POST http://192.168.57.101:60000/_replicate 200
PUT 1000 docs to http://192.168.57.101:60000/ls_db1/ with prefix ls_db1


――――――――――――――――――――――――――――――――――――――――――― test_replication_with_multiple_client_dbs_and_single_sync_gateway_db ――――――――――――――――――――――――――――――――――――――――――――

setup_client_syncgateway_test = {'cluster_config': 'resources/cluster_configs/base_cc', 'ls_url': 'http://192.168.57.101:60000', 'sg_admin_url': 'http://192.168.33.11:4985', 'sg_mode': 'cc', ...}

    @pytest.mark.sanity
    @pytest.mark.listener
    @pytest.mark.syncgateway
    @pytest.mark.replication
    def test_replication_with_multiple_client_dbs_and_single_sync_gateway_db(setup_client_syncgateway_test):
        """Test replication from multiple client dbs to one sync_gateway db"""
    
        cluster_config = setup_client_syncgateway_test["cluster_config"]
        sg_mode = setup_client_syncgateway_test["sg_mode"]
        ls_url = setup_client_syncgateway_test["ls_url"]
        sg_url = setup_client_syncgateway_test["sg_url"]
        sg_admin_url = setup_client_syncgateway_test["sg_admin_url"]
    
        num_docs = 1000
    
        sg_config = sync_gateway_config_path_for_mode("listener_tests/listener_tests", sg_mode)
        c = cluster.Cluster(config=cluster_config)
        c.reset(sg_config_path=sg_config)
    
        log_info("Running 'test_replication_with_multiple_client_dbs_and_single_sync_gateway_db'")
        log_info("ls_url: {}".format(ls_url))
        log_info("sg_admin_url: {}".format(sg_admin_url))
        log_info("sg_url: {}".format(sg_url))
    
        client = MobileRestClient()
    
        sg_db = "db"
        ls_db1 = client.create_database(url=ls_url, name="ls_db1")
        ls_db2 = client.create_database(url=ls_url, name="ls_db2")
    
        # Setup continuous push / pull replication from ls_db1 to sg_db
        client.start_replication(
            url=ls_url,
            continuous=True,
            from_db=ls_db1,
            to_url=sg_admin_url, to_db=sg_db
        )
    
        client.start_replication(
            url=ls_url,
            continuous=True,
            from_url=sg_admin_url, from_db=sg_db,
            to_db=ls_db1
        )
    
        # Setup continuous push / pull replication from ls_db2 to sg_db
        client.start_replication(
            url=ls_url,
            continuous=True,
            from_db=ls_db2,
            to_url=sg_admin_url, to_db=sg_db
        )
    
        client.start_replication(
            url=ls_url,
            continuous=True,
            from_url=sg_admin_url, from_db=sg_db,
            to_db=ls_db2
        )
    
>       ls_db_one_docs = client.add_docs(url=ls_url, db=ls_db1, number=num_docs, id_prefix=ls_db1)

testsuites/listener/shared/client_sg/test_replication.py:949: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
keywords/MobileRestClient.py:1111: in add_docs
    resp = self._session.put("{}/{}/{}".format(url, db, doc_id), data=data)
venv/lib/python2.7/site-packages/requests/sessions.py:521: in put
    return self.request('PUT', url, data=data, **kwargs)
venv/lib/python2.7/site-packages/requests/sessions.py:468: in request
    resp = self.send(prep, **send_kwargs)
venv/lib/python2.7/site-packages/requests/sessions.py:576: in send
    r = adapter.send(request, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <requests.adapters.HTTPAdapter object at 0x10c78b710>, request = <PreparedRequest [PUT]>, stream = False
timeout = <requests.packages.urllib3.util.timeout.Timeout object at 0x10c6991d0>, verify = True, cert = None, proxies = OrderedDict()

    def send(self, request, stream=False, timeout=None, verify=True, cert=None, proxies=None):
        """Sends PreparedRequest object. Returns Response object.
    
            :param request: The :class:`PreparedRequest <PreparedRequest>` being sent.
            :param stream: (optional) Whether to stream the request content.
            :param timeout: (optional) How long to wait for the server to send
                data before giving up, as a float, or a :ref:`(connect timeout,
                read timeout) <timeouts>` tuple.
            :type timeout: float or tuple
            :param verify: (optional) Whether to verify SSL certificates.
            :param cert: (optional) Any user-provided SSL certificate to be trusted.
            :param proxies: (optional) The proxies dictionary to apply to the request.
            """
    
        conn = self.get_connection(request.url, proxies)
    
        self.cert_verify(conn, request.url, verify, cert)
        url = self.request_url(request, proxies)
        self.add_headers(request)
    
        chunked = not (request.body is None or 'Content-Length' in request.headers)
    
        if isinstance(timeout, tuple):
            try:
                connect, read = timeout
                timeout = TimeoutSauce(connect=connect, read=read)
            except ValueError as e:
                # this may raise a string formatting error.
                err = ("Invalid timeout {0}. Pass a (connect, read) "
                       "timeout tuple, or a single float to set "
                       "both timeouts to the same value".format(timeout))
                raise ValueError(err)
        else:
            timeout = TimeoutSauce(connect=timeout, read=timeout)
    
        try:
            if not chunked:
                resp = conn.urlopen(
                    method=request.method,
                    url=url,
                    body=request.body,
                    headers=request.headers,
                    redirect=False,
                    assert_same_host=False,
                    preload_content=False,
                    decode_content=False,
                    retries=self.max_retries,
                    timeout=timeout
                )
    
            # Send the request.
            else:
                if hasattr(conn, 'proxy_pool'):
                    conn = conn.proxy_pool
    
                low_conn = conn._get_conn(timeout=DEFAULT_POOL_TIMEOUT)
    
                try:
                    low_conn.putrequest(request.method,
                                        url,
                                        skip_accept_encoding=True)
    
                    for header, value in request.headers.items():
                        low_conn.putheader(header, value)
    
                    low_conn.endheaders()
    
                    for i in request.body:
                        low_conn.send(hex(len(i))[2:].encode('utf-8'))
                        low_conn.send(b'\r\n')
                        low_conn.send(i)
                        low_conn.send(b'\r\n')
                    low_conn.send(b'0\r\n\r\n')
    
                    # Receive the response from the server
                    try:
                        # For Python 2.7+ versions, use buffering of HTTP
                        # responses
                        r = low_conn.getresponse(buffering=True)
                    except TypeError:
                        # For compatibility with Python 2.6 versions and back
                        r = low_conn.getresponse()
    
                    resp = HTTPResponse.from_httplib(
                        r,
                        pool=conn,
                        connection=low_conn,
                        preload_content=False,
                        decode_content=False
                    )
                except:
                    # If we hit any problems here, clean up the connection.
                    # Then, reraise so that we can handle the actual exception.
                    low_conn.close()
                    raise
    
        except (ProtocolError, socket.error) as err:
>           raise ConnectionError(err, request=request)
E           ConnectionError: ('Connection aborted.', error(54, 'Connection reset by peer'))

venv/lib/python2.7/site-packages/requests/adapters.py:426: ConnectionError

Tearing down test
GET http://192.168.57.101:60000/_all_dbs 200
DELETE http://192.168.57.101:60000/ls_db1 200
DELETE http://192.168.57.101:60000/ls_db2 200
GET http://192.168.57.101:60000/_all_dbs 200
Stopping LiteServ: http://192.168.57.101:60000
```

![screen shot 2017-02-09 at 1 52 11 pm](https://cloud.githubusercontent.com/assets/15889/22804682/f91c7894-eece-11e6-8d8c-abfecb45a4e6.png)

After `/ls_db1/ls_db1_289` request, next request is `/_all_dbs`. This indicates servlet did not recevie the the request of `/ls_db1/ls_db1_290`
```
D/Listener( 2832): authHeader is null
V/Listener( 2832): Not enforcing basic auth -- allowedCredentials null or empty
I/Listener( 2832): [REQUEST] (Acme.Utils.ThreadPool(2)-PooledThread: Acme.Serve.Serve$ServeConnection@5370aa64) HTTP/1.1: http://192.168.57.101:60000/ls_db1/ls_db1_289
V/Listener( 2832): remoteURL() addr=192.168.57.1
V/Listener( 2832): RouterCallbackBlock.onResponseReady() START
V/Listener( 2832): RouterCallbackBlock.onResponseReady() END
V/Listener( 2832): CountDownLatch.await() START
V/Listener( 2832): CountDownLatch.await() END
D/Listener( 2832): authHeader is null
V/Listener( 2832): Not enforcing basic auth -- allowedCredentials null or empty
I/Listener( 2832): [REQUEST] (Acme.Utils.ThreadPool(1)-PooledThread: Acme.Serve.Serve$ServeConnection@5367f600) HTTP/1.1: http://192.168.57.101:60000/_all_dbs
V/Listener( 2832): remoteURL() addr=192.168.57.1
V/Listener( 2832): RouterCallbackBlock.onResponseReady() START
V/Listener( 2832): RouterCallbackBlock.onResponseReady() END
V/Listener( 2832): CountDownLatch.await() START
V/Listener( 2832): CountDownLatch.await() END
D/Listener( 2832): authHeader is null
V/Listener( 2832): Not enforcing basic auth -- allowedCredentials null or empty
I/Listener( 2832): [REQUEST] (Acme.Utils.ThreadPool(1)-PooledThread: Acme.Serve.Serve$ServeConnection@5367f600) HTTP/1.1: http://192.168.57.101:60000/ls_db1
V/Listener( 2832): remoteURL() addr=192.168.57.1
V/Listener( 2832): RouterCallbackBlock.onResponseReady() START
V/Listener( 2832): RouterCallbackBlock.onResponseReady() END
V/Listener( 2832): CountDownLatch.await() START
V/Listener( 2832): CountDownLatch.await() END
D/Listener( 2832): authHeader is null
V/Listener( 2832): Not enforcing basic auth -- allowedCredentials null or empty
I/Listener( 2832): [REQUEST] (Acme.Utils.ThreadPool(1)-PooledThread: Acme.Serve.Serve$ServeConnection@5367f600) HTTP/1.1: http://192.168.57.101:60000/ls_db2
V/Listener( 2832): remoteURL() addr=192.168.57.1
V/Listener( 2832): RouterCallbackBlock.onResponseReady() START
V/Listener( 2832): RouterCallbackBlock.onResponseReady() END
V/Listener( 2832): CountDownLatch.await() START
V/Listener( 2832): CountDownLatch.await() END
D/Listener( 2832): authHeader is null
V/Listener( 2832): Not enforcing basic auth -- allowedCredentials null or empty
I/Listener( 2832): [REQUEST] (Acme.Utils.ThreadPool(1)-PooledThread: Acme.Serve.Serve$ServeConnection@5367f600) HTTP/1.1: http://192.168.57.101:60000/_all_dbs
V/Listener( 2832): remoteURL() addr=192.168.57.1
V/Listener( 2832): RouterCallbackBlock.onResponseReady() START
V/Listener( 2832): RouterCallbackBlock.onResponseReady() END
V/Listener( 2832): CountDownLatch.await() START
V/Listener( 2832): CountDownLatch.await() END
```

![screen shot 2017-02-09 at 3 17 51 pm](https://cloud.githubusercontent.com/assets/15889/22807274/fd93df6e-eeda-11e6-8545-a0afe8a8083c.png)

```
>       ls_db_two_docs = client.add_docs(url=ls_url, db=ls_db2, number=num_docs, id_prefix=ls_db2)

testsuites/listener/shared/client_sg/test_replication.py:952: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
keywords/MobileRestClient.py:1111: in add_docs
    resp = self._session.put("{}/{}/{}".format(url, db, doc_id), data=data)
venv/lib/python2.7/site-packages/requests/sessions.py:521: in put
    return self.request('PUT', url, data=data, **kwargs)
venv/lib/python2.7/site-packages/requests/sessions.py:468: in request
    resp = self.send(prep, **send_kwargs)
venv/lib/python2.7/site-packages/requests/sessions.py:576: in send
    r = adapter.send(request, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <requests.adapters.HTTPAdapter object at 0x109bda290>, request = <PreparedRequest [PUT]>, stream = False
timeout = <requests.packages.urllib3.util.timeout.Timeout object at 0x109b1ee50>, verify = True, cert = None, proxies = OrderedDict()

    def send(self, request, stream=False, timeout=None, verify=True, cert=None, proxies=None):
        """Sends PreparedRequest object. Returns Response object.
    
            :param request: The :class:`PreparedRequest <PreparedRequest>` being sent.
            :param stream: (optional) Whether to stream the request content.
            :param timeout: (optional) How long to wait for the server to send
                data before giving up, as a float, or a :ref:`(connect timeout,
                read timeout) <timeouts>` tuple.
            :type timeout: float or tuple
            :param verify: (optional) Whether to verify SSL certificates.
            :param cert: (optional) Any user-provided SSL certificate to be trusted.
            :param proxies: (optional) The proxies dictionary to apply to the request.
            """
    
        conn = self.get_connection(request.url, proxies)
    
        self.cert_verify(conn, request.url, verify, cert)
        url = self.request_url(request, proxies)
        self.add_headers(request)
    
        chunked = not (request.body is None or 'Content-Length' in request.headers)
    
        if isinstance(timeout, tuple):
            try:
                connect, read = timeout
                timeout = TimeoutSauce(connect=connect, read=read)
            except ValueError as e:
                # this may raise a string formatting error.
                err = ("Invalid timeout {0}. Pass a (connect, read) "
                       "timeout tuple, or a single float to set "
                       "both timeouts to the same value".format(timeout))
                raise ValueError(err)
        else:
            timeout = TimeoutSauce(connect=timeout, read=timeout)
    
        try:
            if not chunked:
                resp = conn.urlopen(
                    method=request.method,
                    url=url,
                    body=request.body,
                    headers=request.headers,
                    redirect=False,
                    assert_same_host=False,
                    preload_content=False,
                    decode_content=False,
                    retries=self.max_retries,
                    timeout=timeout
                )
    
            # Send the request.
            else:
                if hasattr(conn, 'proxy_pool'):
                    conn = conn.proxy_pool
    
                low_conn = conn._get_conn(timeout=DEFAULT_POOL_TIMEOUT)
    
                try:
                    low_conn.putrequest(request.method,
                                        url,
                                        skip_accept_encoding=True)
    
                    for header, value in request.headers.items():
                        low_conn.putheader(header, value)
    
                    low_conn.endheaders()
    
                    for i in request.body:
                        low_conn.send(hex(len(i))[2:].encode('utf-8'))
                        low_conn.send(b'\r\n')
                        low_conn.send(i)
                        low_conn.send(b'\r\n')
                    low_conn.send(b'0\r\n\r\n')
    
                    # Receive the response from the server
                    try:
                        # For Python 2.7+ versions, use buffering of HTTP
                        # responses
                        r = low_conn.getresponse(buffering=True)
                    except TypeError:
                        # For compatibility with Python 2.6 versions and back
                        r = low_conn.getresponse()
    
                    resp = HTTPResponse.from_httplib(
                        r,
                        pool=conn,
                        connection=low_conn,
                        preload_content=False,
                        decode_content=False
                    )
                except:
                    # If we hit any problems here, clean up the connection.
                    # Then, reraise so that we can handle the actual exception.
                    low_conn.close()
                    raise
    
        except (ProtocolError, socket.error) as err:
>           raise ConnectionError(err, request=request)
E           ConnectionError: ('Connection aborted.', BadStatusLine("''",))

venv/lib/python2.7/site-packages/requests/adapters.py:426: ConnectionError

Tearing down test
GET http://192.168.57.101:60000/_all_dbs 200
DELETE http://192.168.57.101:60000/ls_db1 200
DELETE http://192.168.57.101:60000/ls_db2 200
GET http://192.168.57.101:60000/_all_dbs 200
Stopping LiteServ: http://192.168.57.101:60000

Success


      TEARDOWN F setup_client_syncgateway_testTearing down suite ...

TEARDOWN S setup_client_syncgateway_suite
 testsuites/listener/shared/client_sg/test_replication.pytest_replication_with_multiple_client_dbs_and_single_sync_gateway_db ⨯               100% ██████████
---------------------------------------- generated xml file: /Users/hideki/github/mobile-testkit/results/results.xml ----------------------------------------
--------------------------------------- generated html file: /Users/hideki/github/mobile-testkit/results/report.html ----------------------------------------
==================================================================== 17 tests deselected ====================================================================

Results (53.06s):
       1 failed
         - /Users/hideki/github/mobile-testkit/venv/lib/python2.7/site-packages/requests/adapters.py:426: ConnectionError: ('Connection aborted.', BadStatusLine("''",))
      17 deselected
(venv) hideki@hidekis:~/github/mobile-testkit$ 
```

PyTest freezes during TearDown.

```
POST http://192.168.57.102:60000/ls_db1/_all_docs 200
Num found docs: 1963
Num missing docs: 37
POST http://192.168.57.102:60000/ls_db1/_all_docs 200
Num found docs: 1963
Num missing docs: 37


―――――――――――――――――――――― test_replication_with_multiple_client_dbs_and_single_sync_gateway_db ―――――――――――――――――――――――

setup_client_syncgateway_test = {'cluster_config': 'resources/cluster_configs/base_cc', 'ls_url': 'http://192.168.57.102:60000', 'sg_admin_url': 'http://192.168.33.11:4985', 'sg_mode': 'cc', ...}

    @pytest.mark.sanity
    @pytest.mark.listener
    @pytest.mark.syncgateway
    @pytest.mark.replication
    def test_replication_with_multiple_client_dbs_and_single_sync_gateway_db(setup_client_syncgateway_test):
        """Test replication from multiple client dbs to one sync_gateway db"""
    
        cluster_config = setup_client_syncgateway_test["cluster_config"]
        sg_mode = setup_client_syncgateway_test["sg_mode"]
        ls_url = setup_client_syncgateway_test["ls_url"]
        sg_url = setup_client_syncgateway_test["sg_url"]
        sg_admin_url = setup_client_syncgateway_test["sg_admin_url"]
    
        num_docs = 1000
    
        sg_config = sync_gateway_config_path_for_mode("listener_tests/listener_tests", sg_mode)
        c = cluster.Cluster(config=cluster_config)
        c.reset(sg_config_path=sg_config)
    
        log_info("Running 'test_replication_with_multiple_client_dbs_and_single_sync_gateway_db'")
        log_info("ls_url: {}".format(ls_url))
        log_info("sg_admin_url: {}".format(sg_admin_url))
        log_info("sg_url: {}".format(sg_url))
    
        client = MobileRestClient()
    
        sg_db = "db"
        ls_db1 = client.create_database(url=ls_url, name="ls_db1")
        ls_db2 = client.create_database(url=ls_url, name="ls_db2")
    
        # Setup continuous push / pull replication from ls_db1 to sg_db
        client.start_replication(
            url=ls_url,
            continuous=True,
            from_db=ls_db1,
            to_url=sg_admin_url, to_db=sg_db
        )
    
        client.start_replication(
            url=ls_url,
            continuous=True,
            from_url=sg_admin_url, from_db=sg_db,
            to_db=ls_db1
        )
    
        # Setup continuous push / pull replication from ls_db2 to sg_db
        client.start_replication(
            url=ls_url,
            continuous=True,
            from_db=ls_db2,
            to_url=sg_admin_url, to_db=sg_db
        )
    
        client.start_replication(
            url=ls_url,
            continuous=True,
            from_url=sg_admin_url, from_db=sg_db,
            to_db=ls_db2
        )
    
        ls_db_one_docs = client.add_docs(url=ls_url, db=ls_db1, number=num_docs, id_prefix=ls_db1)
        assert len(ls_db_one_docs) == num_docs
    
        ls_db_two_docs = client.add_docs(url=ls_url, db=ls_db2, number=num_docs, id_prefix=ls_db2)
        assert len(ls_db_two_docs) == num_docs
    
        ls_db1_db2_docs = ls_db_one_docs + ls_db_two_docs
    
>       client.verify_docs_present(url=ls_url, db=ls_db1, expected_docs=ls_db1_db2_docs)

testsuites/listener/shared/client_sg/test_replication.py:957: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <keywords.MobileRestClient.MobileRestClient instance at 0x104ae88c0>, url = 'http://192.168.57.102:60000'
db = 'ls_db1'
expected_docs = [{'id': 'ls_db1_0', 'ok': True, 'rev': '1-6a660dfe715b0eb328bbedefc6d03c6a'}, {'id': 'ls_db1_1', 'ok': True, 'rev': '1...1-53c3b04a5604ba405ec8035ae450f1e6'}, {'id': 'ls_db1_5', 'ok': True, 'rev': '1-75f7c403c509dad34973adda83ef7fb9'}, ...]
auth = None, timeout = 300

    def verify_docs_present(self, url, db, expected_docs, auth=None, timeout=CLIENT_REQUEST_TIMEOUT):
        """
            Verifies the expected docs are present in the database using a polling loop with
            POST _all_docs with Listener and a POST _bulk_get for sync_gateway
    
            expected_docs should be a dict {id: {rev: ""}} or
            a list of {id: {rev: ""}}. If the expected docs are a list, they will be converted to a single map.
            """
    
        auth_type = get_auth_type(auth)
        server_type = self.get_server_type(url)
    
        logging.debug(expected_docs)
    
        if isinstance(expected_docs, list):
            # Create single dictionary for comparison, will also blow up for duplicate docs with the same id
            expected_doc_map = {expected_doc["id"]: expected_doc["rev"] for expected_doc in expected_docs}
        elif isinstance(expected_docs, dict):
            # When expected docs is a single doc
            expected_doc_map = {expected_docs["id"]: expected_docs["rev"]}
        else:
            raise TypeError("Verify Docs Preset expects a list or dict of expected docs")
    
        logging.debug(expected_docs)
    
        log_info("Verify {}/{} has {} docs".format(url, db, len(expected_doc_map)), is_verify=True)
    
        start = time.time()
        while True:
    
            if time.time() - start > timeout:
>               raise keywords.exceptions.TimeoutException("Verify Docs Present: TIMEOUT")
E               TimeoutException: Verify Docs Present: TIMEOUT

keywords/MobileRestClient.py:1440: TimeoutException

Tearing down test
GET http://192.168.57.102:60000/_all_dbs 200
```

thread-dump of LiteServ
```
Full thread dump

"<35> ChangeTracker-http://192.168.33.11:4985/db@831697589320" prio=5 runnable
  java.lang.Thread.State: RUNNABLE
	  at libcore.io.Posix.recvfromBytes(Posix.java:-1)
	  at libcore.io.Posix.recvfrom(Posix.java:141)
	  at libcore.io.BlockGuardOs.recvfrom(BlockGuardOs.java:164)
	  at libcore.io.IoBridge.recvfrom(IoBridge.java:506)
	  at java.net.PlainSocketImpl.read(PlainSocketImpl.java:488)
	  at java.net.PlainSocketImpl.access$000(PlainSocketImpl.java:46)
	  at java.net.PlainSocketImpl$PlainSocketInputStream.read(PlainSocketImpl.java:240)
	  at okio.Okio$2.read(Okio.java:140)
	  at okio.AsyncTimeout$2.read(AsyncTimeout.java:238)
	  at okio.RealBufferedSource.request(RealBufferedSource.java:66)
	  at okio.RealBufferedSource.require(RealBufferedSource.java:59)
	  at okio.RealBufferedSource.readHexadecimalUnsignedLong(RealBufferedSource.java:284)
	  at okhttp3.internal.http.Http1xStream$ChunkedSource.readChunkSize(Http1xStream.java:441)
	  at okhttp3.internal.http.Http1xStream$ChunkedSource.read(Http1xStream.java:422)
	  at okio.RealBufferedSource$1.read(RealBufferedSource.java:386)
	  at java.io.InputStream.read(InputStream.java:162)
	  at java.util.zip.InflaterInputStream.fill(InflaterInputStream.java:194)
	  at java.util.zip.InflaterInputStream.read(InflaterInputStream.java:153)
	  at java.util.zip.GZIPInputStream.read(GZIPInputStream.java:163)
	  at com.fasterxml.jackson.core.json.UTF8StreamJsonParser.loadMore(UTF8StreamJsonParser.java:180)
	  at com.fasterxml.jackson.core.json.UTF8StreamJsonParser._skipWSOrEnd2(UTF8StreamJsonParser.java:2777)
	  at com.fasterxml.jackson.core.json.UTF8StreamJsonParser._skipWSOrEnd(UTF8StreamJsonParser.java:2772)
	  at com.fasterxml.jackson.core.json.UTF8StreamJsonParser.nextToken(UTF8StreamJsonParser.java:652)
	  at com.fasterxml.jackson.databind.deser.std.UntypedObjectDeserializer$Vanilla.deserialize(UntypedObjectDeserializer.java:443)
	  at com.fasterxml.jackson.databind.deser.std.MapDeserializer._readAndBindStringMap(MapDeserializer.java:473)
	  at com.fasterxml.jackson.databind.deser.std.MapDeserializer.deserialize(MapDeserializer.java:330)
	  at com.fasterxml.jackson.databind.deser.std.MapDeserializer.deserialize(MapDeserializer.java:25)
	  at com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:3560)
	  at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:2646)
	  at com.couchbase.lite.replicator.ChangeTracker.runLoop(ChangeTracker.java:371)
	  at com.couchbase.lite.replicator.ChangeTracker.run(ChangeTracker.java:246)
	  at java.lang.Thread.run(Thread.java:841)

"<14> Acme.Utils.ThreadPool(0)-PooledThread: Acme.Serve.Serve$ServeConnection@529550a8@831697725024" daemon prio=5 runnable
  java.lang.Thread.State: RUNNABLE
	  at com.couchbase.cbforest.Database.close(Database.java:-1)
	  at com.couchbase.lite.store.ForestDBStore.close(ForestDBStore.java:178)
	  at com.couchbase.lite.Database.close(Database.java:1519)
	  at com.couchbase.lite.Database.delete(Database.java:404)
	  at com.couchbase.lite.router.Router.do_DELETE_Database(Router.java:1151)
	  at java.lang.reflect.Method.invokeNative(Method.java:-1)
	  at java.lang.reflect.Method.invoke(Method.java:515)
	  at com.couchbase.lite.router.Router.start(Router.java:557)
	  at com.couchbase.lite.listener.LiteServlet.service(LiteServlet.java:141)
	  at javax.servlet.http.HttpServlet.service(HttpServlet.java:802)
	  at Acme.Serve.Serve$ServeConnection.runServlet(Serve.java:2408)
	  at Acme.Serve.Serve$ServeConnection.parseRequest(Serve.java:2330)
	  at Acme.Serve.Serve$ServeConnection.run(Serve.java:2110)
	  at Acme.Utils$ThreadPool$PooledThread.run(Utils.java:1238)
	  at java.lang.Thread.run(Thread.java:841)

"<49> CBLRequestWorker-http://192.168.33.11:4985/db-push-e066c-4@831698362768" prio=5 waiting
  java.lang.Thread.State: WAITING
	  at java.lang.Object.wait(Object.java:-1)
	  at java.lang.Thread.parkFor(Thread.java:1205)
	  at sun.misc.Unsafe.park(Unsafe.java:325)
	  at java.util.concurrent.locks.LockSupport.park(LockSupport.java:157)
	  at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2017)
	  at java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:1050)
	  at java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:778)
	  at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1035)
	  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1097)
	  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
	  at java.lang.Thread.run(Thread.java:841)

"<48> CBLRequestWorker-http://192.168.33.11:4985/db-push-e066c-3@831698786048" prio=5 waiting
  java.lang.Thread.State: WAITING
	  at java.lang.Object.wait(Object.java:-1)
	  at java.lang.Thread.parkFor(Thread.java:1205)
	  at sun.misc.Unsafe.park(Unsafe.java:325)
	  at java.util.concurrent.locks.LockSupport.park(LockSupport.java:157)
	  at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2017)
	  at java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:1050)
	  at java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:778)
	  at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1035)
	  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1097)
	  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
	  at java.lang.Thread.run(Thread.java:841)

"<47> CBLRequestWorker-http://192.168.33.11:4985/db-push-e066c-2@831698738752" prio=5 waiting
  java.lang.Thread.State: WAITING
	  at java.lang.Object.wait(Object.java:-1)
	  at java.lang.Thread.parkFor(Thread.java:1205)
	  at sun.misc.Unsafe.park(Unsafe.java:325)
	  at java.util.concurrent.locks.LockSupport.park(LockSupport.java:157)
	  at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2017)
	  at java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:1050)
	  at java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:778)
	  at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1035)
	  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1097)
	  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
	  at java.lang.Thread.run(Thread.java:841)

"<45> CBLRequestWorker-http://192.168.33.11:4985/db-push-e066c-1@831698328576" prio=5 waiting
  java.lang.Thread.State: WAITING
	  at java.lang.Object.wait(Object.java:-1)
	  at java.lang.Thread.parkFor(Thread.java:1205)
	  at sun.misc.Unsafe.park(Unsafe.java:325)
	  at java.util.concurrent.locks.LockSupport.park(LockSupport.java:157)
	  at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2017)
	  at java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:1050)
	  at java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:778)
	  at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1035)
	  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1097)
	  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
	  at java.lang.Thread.run(Thread.java:841)

"<44> CBLRequestWorker-http://192.168.33.11:4985/db-pull-fcc53-4@831698080512" prio=5 waiting
  java.lang.Thread.State: WAITING
	  at java.lang.Object.wait(Object.java:-1)
	  at java.lang.Thread.parkFor(Thread.java:1205)
	  at sun.misc.Unsafe.park(Unsafe.java:325)
	  at java.util.concurrent.locks.LockSupport.park(LockSupport.java:157)
	  at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2017)
	  at java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:1050)
	  at java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:778)
	  at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1035)
	  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1097)
	  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
	  at java.lang.Thread.run(Thread.java:841)

"<39> CBLRequestWorker-http://192.168.33.11:4985/db-pull-fcc53-3@831698673840" prio=5 waiting
  java.lang.Thread.State: WAITING
	  at java.lang.Object.wait(Object.java:-1)
	  at java.lang.Thread.parkFor(Thread.java:1205)
	  at sun.misc.Unsafe.park(Unsafe.java:325)
	  at java.util.concurrent.locks.LockSupport.park(LockSupport.java:157)
	  at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2017)
	  at java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:1050)
	  at java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:778)
	  at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1035)
	  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1097)
	  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
	  at java.lang.Thread.run(Thread.java:841)

"<42> CBLRequestWorker-http://192.168.33.11:4985/db-pull-fcc53-2@831698846144" prio=5 waiting
  java.lang.Thread.State: WAITING
	  at java.lang.Object.wait(Object.java:-1)
	  at java.lang.Thread.parkFor(Thread.java:1205)
	  at sun.misc.Unsafe.park(Unsafe.java:325)
	  at java.util.concurrent.locks.LockSupport.park(LockSupport.java:157)
	  at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2017)
	  at java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:1050)
	  at java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:778)
	  at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1035)
	  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1097)
	  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
	  at java.lang.Thread.run(Thread.java:841)

"<41> CBLRequestWorker-http://192.168.33.11:4985/db-pull-fcc53-1@831698339296" prio=5 waiting
  java.lang.Thread.State: WAITING
	  at java.lang.Object.wait(Object.java:-1)
	  at java.lang.Thread.parkFor(Thread.java:1205)
	  at sun.misc.Unsafe.park(Unsafe.java:325)
	  at java.util.concurrent.locks.LockSupport.park(LockSupport.java:157)
	  at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2017)
	  at java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:1050)
	  at java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:778)
	  at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1035)
	  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1097)
	  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
	  at java.lang.Thread.run(Thread.java:841)

"<38> CBLManagerWorkExecutor@831698412216" prio=5 waiting
  java.lang.Thread.State: WAITING
	  at java.lang.Object.wait(Object.java:-1)
	  at java.lang.Thread.parkFor(Thread.java:1205)
	  at sun.misc.Unsafe.park(Unsafe.java:325)
	  at java.util.concurrent.locks.LockSupport.park(LockSupport.java:157)
	  at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2017)
	  at java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:1050)
	  at java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:778)
	  at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1035)
	  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1097)
	  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
	  at java.lang.Thread.run(Thread.java:841)

"<32> CBLRequestWorker-http://192.168.33.11:4985/db-pull-fcc53-0@831697807712" prio=5 waiting
  java.lang.Thread.State: WAITING
	  at java.lang.Object.wait(Object.java:-1)
	  at java.lang.Thread.parkFor(Thread.java:1205)
	  at sun.misc.Unsafe.park(Unsafe.java:325)
	  at java.util.concurrent.locks.LockSupport.park(LockSupport.java:157)
	  at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2017)
	  at java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:1050)
	  at java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:778)
	  at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1035)
	  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1097)
	  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
	  at java.lang.Thread.run(Thread.java:841)

"<31> CBLReplicationExecutor-http://192.168.33.11:4985/db-pull-fcc53@831698017160" prio=5 waiting
  java.lang.Thread.State: WAITING
	  at java.lang.Object.wait(Object.java:-1)
	  at java.lang.Thread.parkFor(Thread.java:1205)
	  at sun.misc.Unsafe.park(Unsafe.java:325)
	  at java.util.concurrent.locks.LockSupport.park(LockSupport.java:157)
	  at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2017)
	  at java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:1050)
	  at java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:778)
	  at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1035)
	  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1097)
	  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
	  at java.lang.Thread.run(Thread.java:841)

"<28> CBLRequestWorker-http://192.168.33.11:4985/db-push-e066c-0@831697929184" prio=5 waiting
  java.lang.Thread.State: WAITING
	  at java.lang.Object.wait(Object.java:-1)
	  at java.lang.Thread.parkFor(Thread.java:1205)
	  at sun.misc.Unsafe.park(Unsafe.java:325)
	  at java.util.concurrent.locks.LockSupport.park(LockSupport.java:157)
	  at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2017)
	  at java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:1050)
	  at java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:778)
	  at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1035)
	  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1097)
	  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
	  at java.lang.Thread.run(Thread.java:841)

"<26> CBLReplicationExecutor-http://192.168.33.11:4985/db-push-e066c@831697958704" prio=5 waiting
  java.lang.Thread.State: WAITING
	  at java.lang.Object.wait(Object.java:-1)
	  at java.lang.Thread.parkFor(Thread.java:1205)
	  at sun.misc.Unsafe.park(Unsafe.java:325)
	  at java.util.concurrent.locks.LockSupport.park(LockSupport.java:157)
	  at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2017)
	  at java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:1050)
	  at java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:778)
	  at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1035)
	  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1097)
	  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
	  at java.lang.Thread.run(Thread.java:841)

"<43> CBLPusherSupportExecutor-http://192.168.33.11:4985/db@831698693184" prio=5 waiting
  java.lang.Thread.State: WAITING
	  at java.lang.Object.wait(Object.java:-1)
	  at java.lang.Thread.parkFor(Thread.java:1205)
	  at sun.misc.Unsafe.park(Unsafe.java:325)
	  at java.util.concurrent.locks.LockSupport.park(LockSupport.java:157)
	  at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2017)
	  at java.util.concurrent.LinkedBlockingQueue.take(LinkedBlockingQueue.java:410)
	  at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1035)
	  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1097)
	  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
	  at java.lang.Thread.run(Thread.java:841)

"<11> Thread-117@831697438952" prio=5 runnable
  java.lang.Thread.State: RUNNABLE
	  at libcore.io.Posix.accept(Posix.java:-1)
	  at libcore.io.BlockGuardOs.accept(BlockGuardOs.java:55)
	  at java.net.PlainSocketImpl.accept(PlainSocketImpl.java:98)
	  at java.net.ServerSocket.implAccept(ServerSocket.java:203)
	  at java.net.ServerSocket.accept(ServerSocket.java:128)
	  at Acme.Serve.SimpleAcceptor.accept(SimpleAcceptor.java:40)
	  at Acme.Serve.Serve.serve(Serve.java:849)
	  at com.couchbase.lite.listener.LiteServer.serve(LiteServer.java:63)
	  at com.couchbase.lite.listener.LiteListener.run(LiteListener.java:90)
	  at java.lang.Thread.run(Thread.java:841)

"<1> main@831693941720" prio=5 runnable
  java.lang.Thread.State: RUNNABLE
	  at android.os.MessageQueue.nativePollOnce(MessageQueue.java:-1)
	  at android.os.MessageQueue.next(MessageQueue.java:138)
	  at android.os.Looper.loop(Looper.java:123)
	  at android.app.ActivityThread.main(ActivityThread.java:5001)
	  at java.lang.reflect.Method.invokeNative(Method.java:-1)
	  at java.lang.reflect.Method.invoke(Method.java:515)
	  at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:785)
	  at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:601)
	  at dalvik.system.NativeStart.main(NativeStart.java:-1)

"<15> Acme.Utils.ThreadPool(1)-PooledThread: Acme.Serve.Serve$ServeConnection@528d2f44@831697621096" daemon prio=5 waiting for monitor entry
  java.lang.Thread.State: BLOCKED
	  at com.couchbase.lite.Manager.getDatabase(Manager.java:560)
	  at com.couchbase.lite.router.Router.start(Router.java:462)
	  at com.couchbase.lite.listener.LiteServlet.service(LiteServlet.java:141)
	  at javax.servlet.http.HttpServlet.service(HttpServlet.java:802)
	  at Acme.Serve.Serve$ServeConnection.runServlet(Serve.java:2408)
	  at Acme.Serve.Serve$ServeConnection.parseRequest(Serve.java:2330)
	  at Acme.Serve.Serve$ServeConnection.run(Serve.java:2110)
	  at Acme.Utils$ThreadPool$PooledThread.run(Utils.java:1238)
	  at java.lang.Thread.run(Thread.java:841)

"<7> FinalizerDaemon@831696919752" daemon prio=5 waiting
  java.lang.Thread.State: WAITING
	  at java.lang.Object.wait(Object.java:-1)
	  at java.lang.Object.wait(Object.java:401)
	  at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:102)
	  at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:73)
	  at java.lang.Daemons$FinalizerDaemon.run(Daemons.java:170)
	  at java.lang.Thread.run(Thread.java:841)

"<34> OkHttp ConnectionPool@831698010776" daemon prio=5 sleeping
  java.lang.Thread.State: TIMED_WAITING
	  at java.lang.Object.wait(Object.java:-1)
	  at okhttp3.ConnectionPool$1.run(ConnectionPool.java:66)
	  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112)
	  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
	  at java.lang.Thread.run(Thread.java:841)

"<8> FinalizerWatchdogDaemon@831696920168" daemon prio=5 waiting
  java.lang.Thread.State: WAITING
	  at java.lang.Object.wait(Object.java:-1)
	  at java.lang.Object.wait(Object.java:364)
	  at java.lang.Daemons$FinalizerWatchdogDaemon.waitForObject(Daemons.java:230)
	  at java.lang.Daemons$FinalizerWatchdogDaemon.run(Daemons.java:207)
	  at java.lang.Thread.run(Thread.java:841)

"<16> Acme.Utils.ThreadPool-PooledThread: FREE@831697407512" daemon prio=5 waiting
  java.lang.Thread.State: WAITING
	  at java.lang.Object.wait(Object.java:-1)
	  at java.lang.Object.wait(Object.java:364)
	  at Acme.Utils$ThreadPool$PooledThread.run(Utils.java:1232)
	  at java.lang.Thread.run(Thread.java:841)

"<6> ReferenceQueueDaemon@831696919408" daemon prio=5 waiting
  java.lang.Thread.State: WAITING
	  at java.lang.Object.wait(Object.java:-1)
	  at java.lang.Object.wait(Object.java:364)
	  at java.lang.Daemons$ReferenceQueueDaemon.run(Daemons.java:130)
	  at java.lang.Thread.run(Thread.java:841)

"<24> Okio Watchdog@831697704832" daemon prio=5 sleeping
  java.lang.Thread.State: TIMED_WAITING
	  at java.lang.Object.wait(Object.java:-1)
	  at okio.AsyncTimeout.awaitTimeout(AsyncTimeout.java:350)
	  at okio.AsyncTimeout$Watchdog.run(AsyncTimeout.java:313)

"<10> Binder_2@831696938216" prio=5 runnable
  java.lang.Thread.State: RUNNABLE
	  at dalvik.system.NativeStart.run(NativeStart.java:-1)

"<9> Binder_1@831696933120" prio=5 runnable
  java.lang.Thread.State: RUNNABLE
	  at dalvik.system.NativeStart.run(NativeStart.java:-1)

"<5> Compiler@831696919176" daemon prio=5 waiting
  java.lang.Thread.State: WAITING
	  at dalvik.system.NativeStart.run(NativeStart.java:-1)

"<3> Signal Catcher@831696918704" daemon prio=5 waiting
  java.lang.Thread.State: WAITING
	  at dalvik.system.NativeStart.run(NativeStart.java:-1)

"<2> GC@831696918480" daemon prio=5 waiting
  java.lang.Thread.State: WAITING
	  at dalvik.system.NativeStart.run(NativeStart.java:-1)

"<13> KeepAlive cleaner@831697526160" daemon prio=5 sleeping
  java.lang.Thread.State: TIMED_WAITING
	  at java.lang.VMThread.sleep(VMThread.java:-1)
	  at java.lang.Thread.sleep(Thread.java:1013)
	  at java.lang.Thread.sleep(Thread.java:995)
	  at Acme.Serve.Serve$KeepAliveCleaner.run(Serve.java:1721)

"<12> Session cleaner@831697524840" prio=1 sleeping
  java.lang.Thread.State: TIMED_WAITING
	  at java.lang.VMThread.sleep(VMThread.java:-1)
	  at java.lang.Thread.sleep(Thread.java:1013)
	  at java.lang.Thread.sleep(Thread.java:995)
	  at Acme.Serve.Serve$2.run(Serve.java:793)
	  at java.lang.Thread.run(Thread.java:841)
```

--- a/src/main/java/com/couchbase/lite/replicator/PusherInternal.java
+++ b/src/main/java/com/couchbase/lite/replicator/PusherInternal.java
@@ -749,7 +749,7 @@ public class PusherInternal extends ReplicationInternal implements Database.Chan
                     // Skip revisions that originally came from the database I'm syncing to:
                     URL source = change.getSource();
                     if (source != null && source.toURI().equals(remoteUri))
-                        return;
+                        continue;
                     RevisionInternal rev = change.getAddedRevision();
                     if (rev == null)
                         continue;

